# A La Russe
> A La Russe 알고리즘의 구현
- toc: true 
- badges: true
- comments: true
- categories: [알고리즘]
- permalink: /algorithm/a_la_russe
- exec: binder

<br>

### 1. 알고리즘 개요
A La Russe는 두 정수를 곱하는 알고리즘이다. 그러나 우리가 계산하던 방식과는 약간 다르게 접근한다. 3칸을 만들고, 수를 적은 뒤 한쪽엔 2를 곱해가며, 한쪽엔 2를 나눠가며 계산하는 독특한 방식이다.
<br><br>

### 2. 수도 코드
```c
int aLaRusse(int a, int b){
    "3개의 칸을 만들고 a는 1번, b는 2번칸에 쓴다"
            
    while("1번 값" != 0){
           "1번값이 홀수이면, 2번 값을 3번에 옮겨쓴다"
           "1번값을 2로 나누고, 2번값에 2를 곱한다"
    }
        
    return "3번을 모두 더한 값"
}
```
<br><br>

### 3. 알고리즘 구현

In [1]:
#include<stdio.h>

int aLaRusse(int a, int b){
    int c = 0; 
    // 굳이 3칸 안만들고 a, b는 변수 그대로 이용
    
    while(a != 0){
        if(a % 2 != 0){
            c += b;
        }
        
        a /= 2;
        b *= 2;
        
        printf("%d %d %d\n", a, b, c);
    }
    
    return c;
}

int main(){
    int result = aLaRusse(45, 37);
    printf("\n결과 = %d", result);
    
    return 0;
}

22 74 37
11 148 37
5 296 185
2 592 481
1 1184 481
0 2368 1665

결과 = 1665

<br><br>

### 4. 알고리즘 원리

![img](./img/algorithm/001_01.png)

이 알고리즘은 `"2x2 = 1x4"라는 원리`를 가지고 곱셈을 수행한다. 맨 처음에 45가 홀수이기 때문에 하나 떼놓고, 44x37 + 37을 수행한다. 이 때, 37이 한번 덜 곱해져서 더해줬는데, 이 값이 바로 3번째 칸에 쓰이는 값이다 . 위의 원리를 이용하면 44x37 = 22x74와 동일하고 22x74 = 11x148과 동일하다. 이 원리로 첫번째 수가 1이 될때까지 짝수만들고 나머지 빼고 곱셈을 반복한다. 최종적으로 1x1184이 되고 나머지(37+148+296=481)를 모두 더하면 1665가 된다.
<br><br>

### 5. Reference
> youtube: https://youtu.be/XyZ8bAuanHc